analyzers/tests/SonarAnalyzer.Test/TestCases/NamedPlaceholdersShouldBeUnique.cs

```diff
@@ -0,0 +1,298 @@
+
+            logger.Log(LogLevel.Trace, args: new[] { foo, foo }, message: "Hey {foo} and {foo}"); // Noncompliant
+            //                                                                            ^^^
+
+            // Grammar checks
+            logger.LogInformation("Hey {foo} and {$foo} and {@foo}", foo, bar);
+            //                                     ^^^ {{Message template placeholder 'foo' is not unique.}}
+            //                                                ^^^ @-1 {{Message template placeholder 'foo' is not unique.}}
+
+            logger.LogInformation("Hey {foo} and {foo,42} and {foo:format} and {foo,-42:for_{_mat}", foo, bar);
+            //                                    ^^^ {{Message template placeholder 'foo' is not unique.}}
+            //                                                 ^^^ @-1 {{Message template placeholder 'foo' is not unique.}}
+            //                                                                  ^^^ @-2 {{Message template placeholder 'foo' is not unique.}}
+
+            logger.LogInformation("Hey {_foo} and {_foo42} and {_foo42} and {_foo}", foo, bar);
+            //                                                  ^^^^^^ {{Message template placeholder '_foo42' is not unique.}}
+            //                                                               ^^^^ @-1 {{Message template placeholder '_foo' is not unique.}}
+
+            // Multiline
+            logger.LogDebug(
+                message: "Hey {foo} and {foo}", // Noncompliant
+                //                       ^^^
+                eventId: eventId,
+                exception: ex,
+                args: new[] { foo, foo });
+
+            LoggerExtensions.Log(
+                logger,
+                logLevel: LogLevel.Trace,
+                args: Array.Empty<object>(),
+                message: "Hey {foo} and {foo}"); // Noncompliant
+            //                           ^^^
+
+            logger.LogWarning(@"
+                hey there
+                {foo}
+                and {bar}
+                and again {foo}
+                ", foo, bar, foo);
+            //             ^^^ @-1 {{Message template placeholder 'foo' is not unique.}}
+        }
+
+        // Real
+        class MyLogger : ILogger
+        {
+            public IDisposable BeginScope<TState>(TState state) => null;
+            public bool IsEnabled(LogLevel logLevel) => true;
+            public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter) { }
+        }
+
+        // Fake
+        public class NotLogger
+        {
+            public void LogDebug(string message, params object[] args) { }
+        }
+    }
+
+    public static class NotLoggerExtensions
+    {
+        public static void LogCritical(this Program.NotLogger logger, string message, params object[] args) { }
+    }
+}
+
+namespace SerilogTests
```

Comment:
On serilog and NLog I test mostly that I match the signature and argument of the invocation, since the string-oriented checks are the same all across the frameworks, so I put them only in the `Microsoft.Extensions.Loggin` tests.